Resolving bug on CNOS Image download on multiple devices.

amuraleedhar opened the pull request:
##### SUMMARY
This is a fix for the failure of image download on to multiple devices. Error string are popping up on waiting for image transfer from sftp server. This errors are isolated and issue is getting mitigated

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/module_utils/cnos.py

##### ANSIBLE VERSION
ansible 2.4.0.0
  config file = None
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/dist-packages/ansible-2.4.0.0-py2.7.egg/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]

##### ADDITIONAL INFORMATION
If the devices are contacting a sftp server for image for the first time, it will ask for prompts to save this hosts among the known hosts, This will get saved and in next image download server wont ask it, So the logic written for first logic will wait for the prompt and time out. This will get interpreted as Error. With this fix I mitigate this

Commit:
Resolving bug on CNOS Image download on multiple devices.

gundalow commented:
Hi,
We are very close to to releasing `2.4.0`, so I think this is a `2.4.1` candidate. What are your thoughts?

gundalow commented:
bot_status

amuraleedhar commented:
If its not too late, I would still like to have this merged for 2.4.0. And this is just a minor change which just fixes a first time case of image download using an sftp server.

dkasberg commented:
Hi John, it would be appreciated if this could be done in 2.4.0, but understand release logistics.  As Anil mentioned this is a minor change to our module therefore is low risk.

abadger commented:
To an outsider to this code, the risk here is somewhat high.  the code is talking to a service that's outside of our code's control therefore it's hard for the outsider to evaluate if the cornercases in what could be returned would be made better or worse by this change (or if there are no other cornercases).

What's the reward?  Are cnos modules not functional after the initial use?  Is it one specific functionality that is non-functional after initial use?

amuraleedhar commented:
The change in the code is pertaining to image download. The code insertion has been done in waitForDeviceResponse method of cnos.py.  The bug scenario is like this. When trying to download image from a sftp server, using our module cnos_image.py, the image download succeeds but playbook logs indicate it as failed. The functionality is achieved but the playbook status is shown it as failed. The fix is to mitigate that. The image transfer and subsequent installation may take more than expected and in such cases though the image download succeeds, but playbook comes out stating Device timed out.  Playbook is waiting for a response from Device which I am by passing. Then playbook also indicates succeeded and actualy image download also succeeds. This has been tested and perfected in my lab set up. Lenovo Qa has raised this bug and the whole process of QA is followed in fixing this bug. Once the internal Qa approves the fix I am raising this PR in Ansible. I request you to take it forward merging.

amuraleedhar commented:
I am pasting the screen data pertaining to this module run in my test set up as proof
Three devices are getting image updated in paralell using Ansible
<10.240.160.151> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1505460109.61-273205885698052/cnos_image.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1505460109.61-273205885698052/" > /dev/null 2>&1 && sleep 0'
changed: [10.240.160.69] => (item={u'serverpassword': u'Lab4man1', u'imgtype': u'os', u'protocol': u'sftp', u'serverip': u'10.240.160.201', u'serverusername': u'tacacs'}) => {
    "changed": true,
    "failed": false,
    "invocation": {
        "module_args": {
            "deviceType": "NE2572_cnos",
            "enablePassword": null,
            "host": "10.240.160.69",
            "imgpath": "/home/tacacs/NE2572-CNOS-10.6.0.33.img",
            "imgtype": "os",
            "outputfile": "./results/cnos_image_10.240.160.69_output.txt",
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "protocol": "sftp",
            "serverip": "10.240.160.201",
            "serverpassword": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "serverusername": "tacacs",
            "username": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER"
        }
    },
    "item": {
        "imgtype": "os",
        "protocol": "sftp",
        "serverip": "10.240.160.201",
        "serverpassword": "Lab4man1",
        "serverusername": "tacacs"
    },
    "msg": "Image file tranferred to device"
}
changed: [10.240.160.161] => (item={u'serverpassword': u'Lab4man1', u'imgtype': u'os', u'protocol': u'sftp', u'serverip': u'10.240.160.201', u'serverusername': u'tacacs'}) => {
    "changed": true,
    "failed": false,
    "invocation": {
        "module_args": {
            "deviceType": "g8296_cnos",
            "enablePassword": null,
            "host": "10.240.160.161",
            "imgpath": "/home/tacacs/G8296-CNOS-10.6.0.35.img",
            "imgtype": "os",
            "outputfile": "./results/cnos_image_10.240.160.161_output.txt",
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "protocol": "sftp",
            "serverip": "10.240.160.201",
            "serverpassword": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "serverusername": "tacacs",
            "username": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER"
        }
    },
    "item": {
        "imgtype": "os",
        "protocol": "sftp",
        "serverip": "10.240.160.201",
        "serverpassword": "Lab4man1",
        "serverusername": "tacacs"
    },
    "msg": "Image file tranferred to device"
}
changed: [10.240.160.151] => (item={u'serverpassword': u'Lab4man1', u'imgtype': u'os', u'protocol': u'sftp', u'serverip': u'10.240.160.201', u'serverusername': u'tacacs'}) => {
    "changed": true,
    "failed": false,
    "invocation": {
        "module_args": {
            "deviceType": "g8272_cnos",
            "enablePassword": null,
            "host": "10.240.160.151",
            "imgpath": "/home/tacacs/G8272-CNOS-10.6.0.35.img",
            "imgtype": "os",
            "outputfile": "./results/cnos_image_10.240.160.151_output.txt",
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "protocol": "sftp",
            "serverip": "10.240.160.201",
            "serverpassword": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "serverusername": "tacacs",
            "username": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER"
        }
    },
    "item": {
        "imgtype": "os",
        "protocol": "sftp",
        "serverip": "10.240.160.201",
        "serverpassword": "Lab4man1",
        "serverusername": "tacacs"
    },
    "msg": "Image file tranferred to device"
}
META: ran handlers
META: ran handlers

PLAY RECAP *************************************************************************************************************************************************************
10.240.160.151             : ok=1    changed=1    unreachable=0    failed=0
10.240.160.161             : ok=1    changed=1    unreachable=0    failed=0
10.240.160.69              : ok=1    changed=1    unreachable=0    failed=0

amuraleedhar commented:
Adding ansible.log for your purview.

[ansible.zip](https://github.com/ansible/ansible/files/1308256/ansible.zip)

abadger commented:
I've cherry-picked this for 2.4.0rc4.  It sounds like this is a somewhat minor bug to be changing code this late in the rc process so in future releases, we'd probably slip it to the next release.  In this one, however, I already allowed in NXOS changes of a similar magnitude for rc4 because that team was counting on several more rc's being available to make changes in.  So I can't justify to myself leaving this one out just because we hadn't discussed it earlier in the week.

Do remember, though, that in future releases, we probably wouldn't take this change for a prerelease this close to the final release.  Thanks!

amuraleedhar commented:
Thank you.
